src/antsibull/pypi.py

```diff
@@ -0,0 +1,144 @@
+
+"""
+Work with the PyPI API
+"""
+
+import datetime
+from pathlib import Path
+from typing import NamedTuple, Optional
+from urllib.parse import urljoin
+
+import aiohttp
+import pydantic as p
+from antsibull_core.utils.hashing import verify_hash
+
+PYPI_BASE_URL = "https://pypi.org/pypi/"
+
+
+class SdistAndWheel(NamedTuple):
+    """
+    Tuple of UrlInfo objects for a project's sdist and wheel
+    """
+
+    sdist: "UrlInfo"
+    wheel: "UrlInfo"
+
+
+# pyre-ignore[13]: BaseModel initializes attributes when data is loaded
+class Release(p.BaseModel):
+    """
+    Model representing the response from the PyPI API releases endpoint.
+
+    https://warehouse.pypa.io/api-reference/json.html#get--pypi--project_name---version--json
+    """
+
+    info: "ReleaseInfo"
+    last_serial: int
+    urls: list["UrlInfo"]
+    vulnerabilities: list[dict]
+
+    def get_sdist_and_wheel(self) -> SdistAndWheel:
+        sdist: Optional[UrlInfo] = None
+        wheel: Optional[UrlInfo] = None
+        for release in self.urls:
+            if not sdist and release.packagetype == "sdist":
+                sdist = release
+            elif not wheel and release.packagetype == "bdist_wheel":
+                wheel = release
+            msg = f"Not {{}} was uploaded for {self.info.name}=={self.info.version}"
+        if not sdist:
+            raise ValueError(msg.format("sdist"))
+        if not wheel:
+            raise ValueError(msg.format("wheel"))
+        return SdistAndWheel(sdist, wheel)
+
+
+# pyre-ignore[13]: BaseModel initializes attributes when data is loaded
+class ReleaseInfo(p.BaseModel):
+    author: str
+    author_email: str
+    classifiers: list[str]
+    description: str
+    description_content_type: str
+    docs_url: Optional[str]
+    home_page: str
+    keywords: str
+    license: str
+    maintainer: str
+    maintainer_email: str
+    name: str
+    package_url: str
+    platform: Optional[str]
+    project_url: str
+    project_urls: dict[str, str]
+    release_url: str
+    requires_dist: list[str]
+    requires_python: str
+    summary: str
+    version: str
+    yanked: bool
+    yanked_reason: Optional[str]
+
+
+# pyre-ignore[13]: BaseModel initializes attributes when data is loaded
+class UrlInfo(p.BaseModel):
+    comment_text: str
+    digests: dict[str, str]
+    downloads: int
+    filename: str
+    md5_digest: str
+    packagetype: str
+    python_version: str
+    requires_python: Optional[str]
+    size: int
+    upload_time: datetime.datetime
+    url: str
+    yanked: bool
+    yanked_reason: Optional[str]
+
+    @property
+    def sha256sum(self) -> str:
+        return self.digests["sha256"]
+
+    async def verify_local_file(self, file: Path) -> bool:
+        """
+        Check if a local file's name and sha256sum matches this release
+        """
+        checks: list[bool] = [
+            await verify_hash(file, self.sha256sum),
+            file.name == self.filename,
```

Comment:
It would be much cheaper to make this a separate check and bail early, before attempting hashing. It's already performed unconditionally regardless of the hash verification result, anyway.

Reply:
That's true. I guess I'll have to get rid of the list of checks thing. I wish there were async lambdas so I could do a list of those instead.

Reply:
A had some utils for async mapping and stuff somewhere in octomachinery. Plus there was some utility lib for func/itertools like async stuff. But that's an overkill here.
Also, frameworks like Trio have better abstractions for this. And some of them inspired stdlib improvements, but I don't think many would be possible to adopt there. Anyway, I like running everything through the anyio shim layer to get the best of two worlds...